onnxruntime/core/session/custom_ops.cc

```diff
@@ -121,304 +131,392 @@ struct OrtShapeInferContext {
-      out[value.size()] = '\0';
-      *size = value.size() + 1;
+    } else {
+      return OrtApis::CreateStatus(OrtErrorCode::ORT_INVALID_ARGUMENT, "Attribute does not exist.");
+    }
+  });
+}
+
+ORT_API_STATUS_IMPL(OrtApis::ShapeInferContext_SetOutputTypeShape, _In_ const OrtShapeInferContext* context,
+                    _In_ size_t index, _In_ const OrtTensorTypeAndShapeInfo* info) {
+  return ExecuteIfCustomOpsApiEnabled([&]() -> OrtStatusPtr {
+    auto status = context->SetOutputTypeShape(index, info);
+    if (status.IsOK()) {
       return nullptr;
-    } else {  // User has provided a buffer that is not large enough
-      *size = value.size() + 1;
-      return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "Result buffer is not large enough");
+    } else {
+      return OrtApis::CreateStatus(static_cast<OrtErrorCode>(status.Code()), status.ErrorMessage().c_str());
     }
-  }
-  return onnxruntime::ToOrtStatus(status);
-  API_IMPL_END
+  });
 }
 
-#ifdef _WIN32
-#pragma warning(push)
-#pragma warning(disable : 28196 6387)
-#endif
+ORT_API_STATUS_IMPL(OrtApis::ReadOpAttr, _In_ const OrtOpAttr* op_attr, _In_ OrtOpAttrType type, _Inout_ void* data,
+                    _In_ size_t len, _Out_ size_t* out) {
+  return ExecuteIfCustomOpsApiEnabled([&]() -> OrtStatusPtr {
+    if (!op_attr) {
+      return OrtApis::CreateStatus(OrtErrorCode::ORT_INVALID_ARGUMENT, "Invalid attribute.");
+    }
 
-ORT_API_STATUS_IMPL(OrtApis::KernelContext_GetGPUComputeStream, _In_ const OrtKernelContext* context, _Outptr_ void** out) {
-  API_IMPL_BEGIN
-  auto* stream = reinterpret_cast<const onnxruntime::OpKernelContext*>(context)->GetComputeStream();
-  if (stream)
-    *out = stream->GetHandle();
-  else
-    *out = nullptr;
-  return nullptr;
-  API_IMPL_END
-};
+    auto attr = reinterpret_cast<const ONNX_NAMESPACE::AttributeProto*>(op_attr);
+    OrtStatusPtr ret = nullptr;
+    *out = 0;
 
-ORT_API_STATUS_IMPL(OrtApis::KernelContext_GetAllocator, _In_ const OrtKernelContext* context, _In_ const OrtMemoryInfo* mem_info, _Outptr_ OrtAllocator** out) {
-  API_IMPL_BEGIN
-  onnxruntime::AllocatorPtr allocator = reinterpret_cast<const onnxruntime::OpKernelContext*>(context)->GetAllocator(mem_info->device);
-  if (!allocator) {
-    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "No requested allocator available");
-  }
-  std::unique_ptr<onnxruntime::OrtAllocatorImplWrappingIAllocator> p = std::make_unique<onnxruntime::OrtAllocatorImplWrappingIAllocator>(std::move(allocator));
-  *out = p.release();
-  return nullptr;
-  API_IMPL_END
-};
+    switch (type) {
+      case OrtOpAttrType::ORT_OP_ATTR_FLOAT: {
+        if (len < sizeof(float)) {
+          ret = OrtApis::CreateStatus(OrtErrorCode::ORT_INVALID_ARGUMENT,
+                                      "Size of data not large enough to hold a float.");
+        } else {
+          if (attr->has_f()) {
+            auto output_f = reinterpret_cast<float*>(data);
```

Comment:
not related to this PR, but do we need to consider whether `data` is correctly aligned before writing something like a float to it?